src/core/lib/iomgr/tcp_client_cfstream.mm

```diff
@@ -0,0 +1,282 @@
+      case kCFStreamEventCanAcceptBytes:
+      default:
+        // Do nothing; handled by endpoint
+        break;
+    }
+
+    if (grpc_tcp_trace.enabled()) {
+      gpr_log(GPR_DEBUG, "CLIENT_CONNECT :%p connect write callback (%p, %lu, %p)", connect, stream,
+              type, clientCallBackInfo);
+    }
+  });
+  CFWriteStreamSetClient(stream, 0, nil, nil);
+}
+
+static void ParseResolvedAddress(const grpc_resolved_address* addr, CFStringRef* host, int* port) {
+  char *hostPort, *hostString, *portString;
+  grpc_sockaddr_to_string(&hostPort, addr, 1);
+  gpr_split_host_port(hostPort, &hostString, &portString);
+  *host = CFStringCreateWithCString(NULL, hostString, kCFStringEncodingUTF8);
+  gpr_free(hostString);
+  gpr_free(portString);
+  gpr_free(hostPort);
+  *port = grpc_sockaddr_get_port(addr);
+}
+
+static void TCPClientConnectImpl(grpc_closure* closure, grpc_endpoint** ep,
+                                 grpc_pollset_set* interestedParties,
+                                 const grpc_channel_args* channelArgs,
+                                 const grpc_resolved_address* resolvedAddr, grpc_millis deadline) {
+  CFStreamTCPConnect* connect;
+
+  connect = (CFStreamTCPConnect*)gpr_zalloc(sizeof(CFStreamTCPConnect));
+  connect->closure = closure;
+  connect->endpoint = ep;
+  connect->addrName = grpc_sockaddr_to_uri(resolvedAddr);
+  // connect->resource_quota = resource_quota;
+  connect->refs = 2;  // One for the connect operation, one for the timer.
+  gpr_mu_init(&connect->mu);
+
+  if (grpc_tcp_trace.enabled()) {
+    gpr_log(GPR_DEBUG, "CLIENT_CONNECT: %s: asynchronously connecting", connect->addrName);
+  }
+
+  grpc_resource_quota* resourceQuota = grpc_resource_quota_create(NULL);
+  if (channelArgs != NULL) {
+    for (size_t i = 0; i < channelArgs->num_args; i++) {
+      if (0 == strcmp(channelArgs->args[i].key, GRPC_ARG_RESOURCE_QUOTA)) {
+        grpc_resource_quota_unref_internal(resourceQuota);
+        resourceQuota = grpc_resource_quota_ref_internal(
+            (grpc_resource_quota*)channelArgs->args[i].value.pointer.p);
+      }
+    }
+  }
+  connect->resourceQuota = resourceQuota;
+
+  CFReadStreamRef readStream;
+  CFWriteStreamRef writeStream;
+
+  CFStringRef host;
+  int port;
+  ParseResolvedAddress(resolvedAddr, &host, &port);
+  CFStreamCreatePairWithSocketToHost(NULL, host, port, &readStream, &writeStream);
+  CFRelease(host);
+  connect->readStream = readStream;
+  connect->writeStream = writeStream;
+  CFStreamClientContext ctx = {0, static_cast<void*>(connect), nil, nil, nil};
+  CFReadStreamSetClient(readStream, kCFStreamEventErrorOccurred | kCFStreamEventOpenCompleted,
+                        ReadCallback, &ctx);
+  CFWriteStreamSetClient(writeStream, kCFStreamEventErrorOccurred | kCFStreamEventOpenCompleted,
+                         WriteCallback, &ctx);
+  CFReadStreamScheduleWithRunLoop(readStream, CFRunLoopGetMain(), kCFRunLoopCommonModes);
```

Comment:
Although you redispatch callbacks immediately, it would be good to move this entirely off the main runloop. Framerate sensitive apps don't want runloop contention just to service the stream.

Reply:
I tend to leave it as a possible enhancement. This requires an extra dedicated thread in the background running the run-loop, which might not be ideal. I did not have much experience on this, but I do hope to understand how bad it would be if we just dispatch a block in this callback.

Reply:
How complex is it to move this work off of the main thread? Is it worth doing the evaluation of the overhead of performing the block dispatch on the main thread or is it easier to just go ahead and refactor to avoid the block dispatch from the main thread? My gut says that the latter is less work.

Reply:
Probably. Let me think about it.